File: app/src/main/java/code/name/monkey/retromusic/providers/interfaces/Repository.java
package code.name.monkey.retromusic.providers.interfaces;

import code.name.monkey.retromusic.model.Album;
import code.name.monkey.retromusic.model.Artist;
import code.name.monkey.retromusic.model.Genre;
import code.name.monkey.retromusic.model.Playlist;
import code.name.monkey.retromusic.model.Song;
import code.name.monkey.retromusic.model.smartplaylist.AbsSmartPlaylist;
import io.reactivex.Observable;
import java.io.File;
import java.util.ArrayList;

/**
 * Created by hemanths on 11/08/17.
 */

public interface Repository {

    Observable<ArrayList<Song>> getAllSongs();

    Observable<ArrayList<AbsSmartPlaylist>> getSuggestionSongs();

    Observable<Song> getSong(int id);

    Observable<ArrayList<Album>> getAllAlbums();

    Observable<ArrayList<Album>> getRecentAlbums();

    Observable<ArrayList<Album>> getTopAlbums();

    Observable<Album> getAlbum(int albumId);

    Observable<ArrayList<Artist>> getAllArtists();

    Observable<ArrayList<Artist>> getRecentArtists();

    Observable<ArrayList<Artist>> getTopArtists();


    Observable<Artist> getArtistById(long artistId);

    Observable<ArrayList<Playlist>> getAllPlaylists();

    Observable<ArrayList<Song>> getFavoriteSongs();

    Observable<ArrayList<Object>> search(String query);

    Observable<ArrayList<Song>> getPlaylistSongs(Playlist playlist);

    Observable<ArrayList<Playlist>> getHomeList();

    Observable<ArrayList<AbsSmartPlaylist>> getAllThings();

    Observable<ArrayList<Genre>> getAllGenres();

    Observable<ArrayList<Song>> getGenre(int genreId);

    Observable<File> downloadLrcFile(final String title, final String artist, final long duration);

}

File: app/src/main/java/code/name/monkey/retromusic/providers/interfaces/Repository.kt
/*
 * Copyright (c) 2019 Hemanth Savarala.
 *
 * Licensed under the GNU General Public License v3
 *
 * This is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by
 *  the Free Software Foundation either version 3 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 */

package code.name.monkey.retromusic.providers.interfaces

import code.name.monkey.retromusic.Result
import code.name.monkey.retromusic.model.*
import code.name.monkey.retromusic.network.model.LastFmAlbum
import code.name.monkey.retromusic.network.model.LastFmArtist
import kotlinx.coroutines.flow.Flow

/**
 * Created by hemanths on 11/08/17.
 */

interface Repository {

    suspend fun allAlbums(): List<Album>

    suspend fun albumById(albumId: Int): Album

    suspend fun allSongs(): List<Song>

    suspend fun allArtists(): List<Artist>

    suspend fun allPlaylists(): List<Playlist>

    suspend fun allGenres(): List<Genre>

    suspend fun search(query: String?): MutableList<Any>

    suspend fun getPlaylistSongs(playlist: Playlist): ArrayList<Song>

    suspend fun getGenre(genreId: Int): ArrayList<Song>

    suspend fun artistInfo(name: String, lang: String?, cache: String?): LastFmArtist

    suspend fun albumInfo(artist: String, album: String): LastFmAlbum

    suspend fun artistById(artistId: Int): Artist

    suspend fun recentArtists(): List<Artist>

    suspend fun topArtists(): List<Artist>

    suspend fun topAlbums(): List<Album>

    suspend fun recentAlbums(): List<Album>

    suspend fun recentArtistsHome(): Home

    suspend fun topArtistsHome(): Home

    suspend fun topAlbumsHome(): Home

    suspend fun recentAlbumsHome(): Home

    suspend fun favoritePlaylistHome(): Home

    suspend fun suggestionsHome(): Home

    suspend fun genresHome(): Home

    suspend fun homeSections(): List<Home>

   suspend fun homeSectionsFlow(): Flow<Result<List<Home>>>

    fun songsFlow(): Flow<Result<List<Song>>>

    fun albumsFlow(): Flow<Result<List<Album>>>

    fun artistsFlow(): Flow<Result<List<Artist>>>

    fun playlistsFlow(): Flow<Result<List<Playlist>>>

    fun genresFlow(): Flow<Result<List<Genre>>>

}